src/lib.rs

```diff
@@ -81,46 +133,54 @@ pub fn multi_index_map(input: proc_macro::TokenStream) -> proc_macro::TokenStrea
+    //          + If there are exactly one index in the container, then the index has to be idx, remove the key from the lookup table
+    let removes: Vec<proc_macro2::TokenStream> = fields_to_index().map(|f| {
+        let field_name = f.ident.as_ref().unwrap();
+        let field_name_string = field_name.to_string();
+        let index_name = format_ident!("_{}_index", field_name);
+        let error_msg = format!("Internal invariants broken, unable to find element in index '{field_name_string}' despite being present in another");
+        let (_ordering, uniqueness) = get_index_kind(f).unwrap_or_else(|| {
+            abort_call_site!("Attributes must be in the style #[multi_index(hashed_unique)]")
+        });
 
-            match uniqueness {
-                Uniqueness::Unique => quote! {
-                    // For unique indexes we know that removing an element will not affect any other elements
-                    let removed_elem = self.#index_name.remove(&elem_orig.#field_name);
-                },
-                Uniqueness::NonUnique => quote! {
-                    // For non-unique indexes we must verify that we have not affected any other elements
-                    if let Some(mut elems) = self.#index_name.remove(&elem_orig.#field_name) {
-                        // If any other elements share the same non-unique index, we must reinsert them into this index
-                        if elems.len() > 1 {
-                            let pos = elems.iter().position(|e| *e == idx).expect(#error_msg);
-                            elems.remove(pos);
-                            self.#index_name.insert(elem_orig.#field_name.clone(), elems);
+        match uniqueness {
+            Uniqueness::Unique => quote! {
+                let _removed_elem = self.#index_name.remove(&elem_orig.#field_name);
+            },
+            Uniqueness::NonUnique => quote! {
+                let key_to_remove = &elem_orig.#field_name;
+                if let Some(elems) = self.#index_name.get_mut(key_to_remove) {
+                    if elems.len() > 1 {
+                        if !elems.remove(&idx){
+                            panic!(#error_msg);
                         }
+                    } else {
+                        self.#index_name.remove(key_to_remove);
                     }
                 }
-            }
 
-        })
-        .collect();
+            }
+        }
+    }).collect();
 
 
-    // For each indexed field generate a TokenStream representing the combined remove and insert from that field's lookup table.
+    // For each indexed field generate a TokenStream representing the combined remove and insert from that field's lookup table. Used in modifier
+    // Run after an element is already modified in the backing storage. The element before the change is stored in `elem_orig`. The element after change is stored in reference `elem` (inside the backing storage). The index of `elem` in the backing storage is `idx`
```

Comment:
these comments are too long, it will be great to keep shorter than 110.

Reply:
[f26ab1c](https://github.com/lun3x/multi_index_map/pull/23/commits/f26ab1c1e757d2a936ab27fc42129bba1adec0eb)